s3/policy/statement_builder.go

```diff
@@ -0,0 +1,56 @@
+package policy
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"log"
+)
+
+type Statement struct {
+	Effect    string    `json:"Effect"`
+	Action    []string  `json:"Action"`
+	Resource  []string  `json:"Resource"`
+	Principal Principal `json:"Principal"`
+}
+
+type Principal struct {
+	AWS string `json:"AWS"`
+}
+
+const (
+	ReadOnlyPermissionName  = "read-only"
+	ReadWritePermissionName = "read-write"
+)
+
+func BuildStatement(bucketName string, iamUser iam.User, permissionName string) Statement {
+	var actions []string
+
+	if permissionName == ReadOnlyPermissionName {
+		actions = []string{
+			"s3:GetBucketLocation",
+			"s3:ListBucket",
+			"s3:GetObject",
+		}
+	} else if permissionName == ReadWritePermissionName {
+		actions = []string{
+			"s3:GetBucketLocation",
+			"s3:ListBucket",
+			"s3:GetObject",
+			"s3:PutObject",
+			"s3:DeleteObject",
+		}
+	} else {
+		log.Panicf("unknown permission name %s", permissionName)
```

Comment:
I think the function signature for this should change to `(Statement, error)` and this should be error.Formatf or an error type, and then the corresponding failure case handled in the client.

Reply:
The theory was that we'd validate this in `validPermissionsRequested`, and by the time we get here it's a programming error if it's not an expected value. Would be nice if the type system were strong enough to enforce this, but y'know, go.

Might be better just to bubble the error, as you say.